nodebuilder/p2p/metrics.go

```diff
@@ -0,0 +1,116 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	rcmgr "github.com/libp2p/go-libp2p/p2p/host/resource-manager"
+	rcmgrObs "github.com/libp2p/go-libp2p/p2p/host/resource-manager/obs"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.uber.org/fx"
+
+	"github.com/celestiaorg/celestia-node/nodebuilder/node"
+)
+
+type MetricsConfig struct {
+	// Prometheus Agent http endpoint configuration
+	PrometheusAgentEndpoint string
+	PrometheusAgentPort     string
+}
+
+// DefaultMetricsConfig returns default configuration for P2P subsystem.
+func DefaultMetricsConfig() MetricsConfig {
+	return MetricsConfig{
+		PrometheusAgentEndpoint: "/metrics",
+		PrometheusAgentPort:     "8890",
+	}
+}
+
+func (cfg *MetricsConfig) Validate() error {
+	if cfg.PrometheusAgentEndpoint == "" {
+		return fmt.Errorf("libp2p metrics: prometheus agent endpoint cannot be empty")
+	}
+	endpointPattern := "^/\\w+$"
+	regex := regexp.MustCompile(endpointPattern)
+	if !regex.MatchString(cfg.PrometheusAgentEndpoint) {
+		return fmt.Errorf("libp2p metrics: prometheus agent endpoint must match %s", "/{endpoint}")
+	}
+
+	if cfg.PrometheusAgentPort == "" {
+		return fmt.Errorf("libp2p metrics: prometheus agent port cannot be empty")
+	}
+	return nil
+}
+
+// WithLibp2pMetrics option sets up native libp2p metrics up
+func WithLibp2pMetrics(lifecycle fx.Lifecycle, cfg Config) error {
+	rcmgrObs.MustRegisterWith(prometheus.DefaultRegisterer)
+
+	reg := prometheus.DefaultRegisterer
+	registry := reg.(*prometheus.Registry)
+
+	mux := http.NewServeMux()
+	handler := promhttp.HandlerFor(registry, promhttp.HandlerOpts{Registry: reg})
+	mux.Handle(cfg.Metrics.PrometheusAgentEndpoint, handler)
+
+	promHTTPServer := &http.Server{
+		Addr:              fmt.Sprintf(":%s", cfg.Metrics.PrometheusAgentPort),
+		Handler:           mux,
+		ReadHeaderTimeout: 10 * time.Second,
+	}
+
+	lifecycle.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			go func() {
+				if err := promHTTPServer.ListenAndServe(); err != nil {
+					log.Error("Error starting Prometheus metrics exporter http server")
+					panic(err)
+				}
+			}()
```

Comment:
We are running yet another HTTP server besides gateway, RPC, pprof. Instead, we should have one to register new endpoints as requested.

Reply:
Nothing needs to be fixed in this PR specifically re this. This is part of a larger issue we have that is out of the scope of this PR. 

Note that fixing this would require us to have a port config for Prometheus metrics, and we could also hardcode the endpoint name. (I would honestly hardcode it already, see 0 value in making it configurable but could be wrong)

Reply:
I haven't made one yet. Anyway, it's too big of a refactor for now to do it